packages/application-shell/src/components/navbar/use-navbar-state-manager.ts

```diff
@@ -0,0 +1,197 @@
+  const navBarNode = React.useRef<HTMLElement>(null);
+  const areProjectExtensionsEnabled = useFeatureToggle(PROJECT_EXTENSIONS);
+  const applicationsNavBarMenu = useApplicationsMenu<'navBar'>('navBar', {
+    queryOptions: {
+      onError: reportErrorToSentry,
+    },
+    skipRemoteQuery: !props.environment.servedByProxy,
+    loadMenuConfig: props.DEV_ONLY__loadNavbarMenuConfig,
+  });
+  const { data: projectExtensionsQuery } = useQuery<
+    TFetchProjectExtensionsNavbarQuery,
+    TFetchProjectExtensionsNavbarQueryVariables
+  >(FetchProjectExtensionsNavbar, {
+    skip: !props.environment.servedByProxy || !areProjectExtensionsEnabled,
+    variables: {
+      target: GRAPHQL_TARGETS.SETTINGS_SERVICE,
+    },
+    fetchPolicy: 'cache-and-network',
+    onError: reportErrorToSentry,
+  });
+  const customAppsMenu =
+    projectExtensionsQuery &&
+    projectExtensionsQuery.projectExtension &&
+    projectExtensionsQuery.projectExtension.applications
+      ? projectExtensionsQuery.projectExtension.applications
+          .map<TNavbarMenu | undefined>(app => {
+            // Map the menu properties to match the one from the proxy schema.
+            // This is to ensure that the menu object is the same from the proxy
+            // config and the custom apps config, thus allowing them to be
+            // concatenated and rendered the same way.
+            if (!app.navbarMenu) return;
+            return {
+              key: app.navbarMenu.key,
+              uriPath: app.navbarMenu.uriPath,
+              labelAllLocales: app.navbarMenu.labelAllLocales || [],
+              icon: app.navbarMenu.icon,
+              featureToggle: app.navbarMenu.featureToggle,
+              permissions: app.navbarMenu.permissions as string[],
+              submenu: (app.navbarMenu.submenu || []).map(menu => ({
+                key: menu.key,
+                uriPath: menu.uriPath,
+                labelAllLocales: menu.labelAllLocales || [],
+                featureToggle: menu.featureToggle,
+                permissions: menu.permissions as string[],
+                menuVisibility: undefined,
+                actionRights: undefined,
+                dataFences: undefined,
+              })),
+              menuVisibility: undefined,
+              actionRights: undefined,
+              dataFences: undefined,
+              shouldRenderDivider: false,
+            };
+          })
+          .filter(nonNullable)
+      : [];
+  const cachedIsForcedMenuOpen = window.localStorage.getItem(
+    STORAGE_KEYS.IS_FORCED_MENU_OPEN
+  );
+  const isForcedMenuOpen =
+    typeof cachedIsForcedMenuOpen === 'string'
+      ? cachedIsForcedMenuOpen === 'true'
+      : null;
+
+  const [activeItemIndex, setActiveItemIndex] = React.useState<string | null>(
+    null
+  );
+  const [isExpanderVisible, setIsExpanderVisible] = React.useState(true);
+  const [isMenuOpen, setIsMenuOpen] = React.useState(false);
```

Comment:
Not sure if I recall correctly but this is "just" a toggle right? Do we have more of these in app-kit so that we could add a `const {on, toggle, setOn, setOff} = useToggle()` maybe? Just ain idea...

Reply:
Hmm yes and now...I refactored the 3 states into a reducer now, as I was noticing some flickering in some edge cases, probably because of having multiple "`setState`" calls.
I think a reducer in this case makes more sense, cause those 3 states are often updated together.

Reply:
Makes sense yes. I often start to think that there are more cases for useReducer than one initially thinks. After all, useState just builds on it.